docs/source/en/agents.md

```diff
@@ -0,0 +1,570 @@
+To initialize an agent, you need these arguments:
+
+- an LLM to power your agent - the agent is not exactly the LLM, it’s more like the agent is a program that uses an LLM as its engine.
+- a system prompt: what the LLM engine will be prompted with to generate its output
+- a toolbox from which the agent pick tools to execute
+- a parser to extract from the LLM output which tools are to call and with which arguments
+
+To start with, please install the `agents` extras in order to install all default dependencies.
+
+```bash
+pip install transformers[agents]
+```
+
+To build your LLM engine, you have to define a `llm_engine` method, that will be given text and return text. This callable needs to accept a `stop` argument defining stop sequences indicating when to stop generating its output. For instance as follows:
+
+```python
+from huggingface_hub import login, InferenceClient
+
+login("<YOUR_HUGGINGFACEHUB_API_TOKEN>")
+
+client = InferenceClient(model="mistralai/Mixtral-8x7B-Instruct-v0.1")
+
+def llm_engine(query: str, stop=["Task"]) -> str:
+    response = client.text_generation(query, stop_sequences=stop, return_full_text=False, max_new_tokens=1000)
+    for stop_seq in stop:
+        if response[-len(stop_seq) :] == stop_seq:
+            response = response[: -len(stop_seq)]
+    return response
+```
+
+You could use any other `llm_engine` method, as long as:
+
+- it takes a `str` as input and returns a `str`
+- it will stop generating output _before_ the sequences passed in argument `stop`
+
+You also need a `tools` argument, as a list of `Tools`. We can start with an empty one but specify `add_base_tools` to start from a default set of arguments.
+
+Then you can define your agent and run it.
+
+```python
+from transformers import CodeAgent
+
+agent = CodeAgent(llm_engine=llm_engine, tools=[], add_base_tools=True)
+
+agent.run("Draw me a picture of rivers and lakes")
+```
+
+The prompt and output parser were automatically defined here, but you can easily inspect them.
+
+![Untitled](Doc%20agents%20ac753b9a3b934eaba22f659ba994c4bd/Untitled%201.png)
+
+```python
+agent.prompt_template
+```
+
+Note that your agent is powered by a LLM, so small variations in your prompt might yield completely different results. It's important to explain as clearly as possible the task you want to perform. We go more in-depth on how to write good prompts [here](custom_tools#writing-good-user-inputs).
+
+Every [`~Agent.run`] operation is independent, so you can run it several times in a row with different tasks.
+
+
+## Implementation of agents
+
+### ReactAgent
+
+This is the go-to agent to solve reasoning tasks, since the ReAct framework makes it really efficient to think on the basis of its previous observations.
+
+### CodeAgent
+
+This class is the one to use when trying to execute multimodal tasks, since it natively handles different input and output types for its tools. It has a planning step, then generates python code to execute all actions at once.
```

Comment:
Thank you! I moved that section up. Since writing the description above I've added another PR that separates ReactAgent into ReactJSONAgent and ReactCodeAgent, so I kept the structure with titles (i.e. separate `### React agents` from `### Code agent`)